### PR TITLE
added {http,https,no} proxy support to docker engines

### PR DESCRIPTION
**Pull request by udryan10**
Adds HTTP_PROXY, HTTPS_PROXY and NO_PROXY support to docker engines per issue #1319

### CODE REVIEW

**Comment by nathanleclaire** [comment]
Hm, cool, thanks for the contribution. 

I'm wondering if it might be better to have just one flag, which will then apply to any environment variable, instead of three (and possibly more in the future) different flags?

That way, you would : 

```console
$ docker-machine create -d provider --engine-env HTTP_PROXY=http://foo.com --engine-env HTTPS_PROXY=bar.com proxybox
```

What do you think?

**Comment by udryan10** [comment]
@nathanleclaire I like it! I'll update the PR.

**Comment by udryan10** [comment]
@nathanleclaire the PR now has code that implements the more general `--engine-env` flag

**Comment by nathanleclaire** [comment]
Very cool.  Any ideas how to test this in the integration suite?

**Comment by udryan10** [comment]
@nathanleclaire I came up with a test that should work well -- spins up a machine and checks `/proc/<docker daemon pid>/environs` for the set Environment Variable. Let me know what you think.

**Comment by ehazlett** [comment]
:+1: for this -- i know a lot will benefit :)

**Comment by nathanleclaire** [comment]
Looking pretty good --- we are about to merge in some structural changes to our docs, I will need you to rebase after that and then we will be ready to rock.  I'll ping you when it's ready to rebase.

**Comment by nathanleclaire** [comment]
OK I've merged that docs PR -- please rebase when it is convenient for you.  Thanks!

**Comment by udryan10** [comment]
@nathanleclaire I have rebased my doc changes into the new doc structure

File: docs/reference/create.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<!--[metadata]>
+## Specifying Docker Swarm options for the created machine<!--[metadata]>
```
Comment (nathanleclaire) [inline docs/reference/create.md]: I'm not sure this is supposed to be here?

**Comment by nathanleclaire** [comment]
cc @ehazlett for review of this PR

**Comment by udryan10** [comment]
@nathanleclaire fixed up the mistake in the `docs/reference/create.md` you referenced - apologies for that oversight.

**Comment by nathanleclaire** [comment]
Thanks @udryan10 and no problem.

Hey @tianon, could you please take a quick look at these changes to the various OS configurations and note if anything strikes you as incorrect?  For instance, I'm not sure if we should do `export VAR=blah` in `/var/lib/boot2docker/profile`, the rest of environment variables set there are not exported.

File: docs/reference/create.md [hunk 2]
```diff
@@ -146,6 +147,17 @@ $ docker-machine create -d virtualbox \
     gdns
 ```
 
+Additionally, Docker Machine supports a flag, `--engine-env`, which can be used to
+specify arbitrary environment variables to be set within the engine with the syntax `--engine-env name=value`. For example, to specify that the engine should use `example.com` as the proxy server, you could run the following create command:
+
+```
+$ docker-machine create -d virtualbox \
+    --engine-env HTTP_PROXY=http://example.com:8080 \
+    --engine-env HTTPS_PROXY=http://example.com:8080 \
```
Comment (nathanleclaire) [inline docs/reference/create.md]: Heh... this is HTTPS_PROXY but the proto specified is `http` ;D

File: docs/reference/create.md [hunk 3]
```diff
@@ -146,6 +147,17 @@ $ docker-machine create -d virtualbox \
...
+    --engine-env NO_PROXY=example2.com \
+    proxbox
```
Comment (nathanleclaire) [inline docs/reference/create.md]: Nice name :)

Comment (udryan10) [inline docs/reference/create.md]: :+1:

**Comment by tianon** [comment]
It all looks reasonably sane, but what happens when I need quotes in my environment variable values?  ie: `--engine-env SOME_VAR='some "value" '` (which will come through into Go as `"SOME_VAR=some \"value\" "`)

**Comment by tianon** [comment]
(ie, we probably ought to pass these through `%q` from `fmt` at the very least)

**Comment by udryan10** [comment]
I've updated the code to run the env's through %q

**Comment by ehazlett** [comment]
LGTM

**Comment by nathanleclaire** [comment]
LGTM

**Comment by chadswen** [comment]
It seems that some aspect of proxy variable parsing can be picky about the proxy URL schema. I was getting "Proxy Authentication" errors with a proxy that doesn't require credentials. Adding a trailing slash to my proxy variables fixed the issue, as in:
```
docker-machine create -d virtualbox \
    --engine-env HTTP_PROXY=http://example.com:8080/ \
    --engine-env HTTPS_PROXY=https://example.com:8080/ \
    default
```
This format is used in the [Docker Engine example](https://docs.docker.com/articles/systemd/#http-proxy), but it's not used in the [`docker-machine create` example](https://github.com/docker/machine/commit/ad6d8d49a8f281a167a689f000045f5bea469a60#diff-e22bac5e382dbfda7cf2840ffff1b3bfR154).

**Comment by nathanleclaire** [comment]
@chadswen Could I get you to file a new issue about the documentation change required and cc @udryan10 and myself please?  Thanks!

**Pull request merged**